### PR TITLE
User-defined restrictions (clean history)

### PR DESCRIPTION
**Pull request by ansharlubis**
The cleaned version of the PR #1048.

### CODE REVIEW

**Comment by ansharlubis** [comment]
Notes from @certik:
In subsequent PRs we can also then provide some explicit functions (add/zero/etc) for integers, and use the general mechanism even for integers, and then remove the special casing for integers from the instantiation code.

File: src/lpython/semantics/python_ast_to_asr.cpp [hunk 1]
```diff
@@ -669,6 +670,49 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+                            std::string msg = "The restriction " + keyword_arg + 
+                                " can't be assigned with a non-function " + arg_id;
+                            throw SemanticError(msg, loc);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    */
```
Comment (certik) [inline src/lpython/semantics/python_ast_to_asr.cpp:714]: What is the purpose of this function? It does not seem to be used anywhere.

Comment (ansharlubis) [reply src/lpython/semantics/python_ast_to_asr.cpp:714]: Sorry, I forgot to remove the function after merging.

**Review by certik — APPROVED** [review]
I reviewed this, locally as well, it's a huge change and I think we can merge it. Some things might have broke, I am not 100% sure, but I think the big issues are done, so let's merge it, and we can fix up later if anything broke.

**Comment by certik** [comment]
I merged it. Thank you @ansharlubis for this amazing work. This lays the foundation to generics. 

@czgdp1807, if you notice anything that broke due to this PR, just report it, and we'll fix it. There might be some small things.

@Oshanath, we can now use this in LFortran, it does the heavy lifting.

**Pull request merged**